onnx/checker.cc

```diff
@@ -221,6 +225,31 @@ void check_tensor(const TensorProto& tensor, const CheckerContext& ctx) {
 #undef check_field
 }
 
+void check_sparse_tensor(
+    const SparseTensorProto& sparse,
+    const CheckerContext& ctx) {
+  int64_t nnz = 0;
+  if (sparse.has_values()) {
+    const TensorProto& values = sparse.values();
+    check_tensor(values, ctx);
+    // TBD: Should we generalize and permit tensors of rank > 1?
```

Comment:
I'd like to have higher rank. Examples:
- In multi-class linear classification, we usually store the linear coefficient vectors for all classes as a 2-D tensor.
- In some training algorithms (see truncated Newton methods as an example), a lot of sparse-matrix-vector produces are needed.
- The output of OneHotEncoder/TFIDF should be a sparse tensor and that sparse tensor's rank can be larger than 1.

Reply:
Hi, this question is not about the rank of the sparse-tensor: we do permit sparse-tensors of arbitrary rank. The question is about whether the value stored for a given sparse-index can be, say, N floats instead of a single float. Pytorch, for example, allows this. It allows us to take a d-dimensional tensor, and say that the first s-dimensions are sparse and the remaining (d-s) dimensions are dense. In this case, the values field becomes a (d-s)-dimensional tensor, while the indices becomes a [NNZ, s] dimensional tensor. Representation-wise, this is easy enough to support. However, the implementation of ops will be somewhat more complex to support this generality (vs. treating each value as a scalar).

Reply:
I got your point. This might be helpful for scientific simulations where their matrices can have block-diagonal structure. However, I don't feel they will be ONNX's major users at this stage. Let's not to allow the mixture of sparse and dense for now (as your design can be extended to cover that in the future).